tests/main/test_tool_messages.py

```diff
@@ -1187,6 +1197,410 @@ def agent_response(self, msg: str | ChatDocument) -> Any:
+                use_tools=not use_fn_api,
+                strict_recovery=True,
+                llm=OpenAIGPTConfig(
+                    supports_json_schema=True,
+                    supports_strict_tools=True,
+                ),
+            )
+        )
+        agent.enable_message(NabroskiTool)
+        agent.enable_message(CoriolisTool)
+        agent.enable_message(EulerTool)
+
+        agent.message_history = [
+            LLMMessage(
+                role=Role.SYSTEM,
+                content="You are a helpful assistant.",
+            ),
+            LLMMessage(
+                role=Role.USER,
+                content="""
+                Please give me an example of a Nabroski, Coriolis, or Euler call.
+                """,
+            ),
+            LLMMessage(
+                role=Role.ASSISTANT,
+                content=attempt if isinstance(attempt, str) else attempt.content,
+                tool_calls=None if isinstance(attempt, str) else attempt.oai_tool_calls,
+                function_call=(
+                    None if isinstance(attempt, str) else attempt.function_call
+                ),
+            ),
+        ]
+        if (
+            use_fn_api
+            and use_tools_api
+            and isinstance(attempt, ChatDocument)
+            and attempt.oai_tool_calls is not None
+        ):
+            agent.message_history.extend(
+                [
+                    LLMMessage(
+                        role=Role.TOOL,
+                        tool_call_id=t.id,
+                        content="error",
+                    )
+                    for t in attempt.oai_tool_calls
+                ]
+            )
+
+        agent.handle_message(attempt)
+        assert agent.tool_error
+        response = agent.llm_response(
+            """
+            There was an error in your attempted tool/function call. Please correct it.
+            """
+        )
+        assert response is not None
+        result = agent.handle_message(response)
+        assert result is not None
+        if isinstance(result, ChatDocument):
+            return result.content
+
+        return result
+
+    def to_attempt(attempt: LLMFunctionCall) -> str | ChatDocument:
+        if not use_fn_api:
+            return json.dumps(
+                {
+                    "request": attempt.name,
+                    **(attempt.arguments or {}),
+                }
+            )
+
+        if use_tools_api:
+            return ChatDocument(
+                content="",
+                metadata=ChatDocMetaData(sender=Entity.LLM),
+                oai_tool_calls=[
+                    OpenAIToolCall(
+                        id="call-1234657",
+                        function=attempt,
+                    )
+                ],
+            )
+
+        return ChatDocument(
+            content="",
+            metadata=ChatDocMetaData(sender=Entity.LLM),
+            function_call=attempt,
+        )
+
+    assert (
```

Comment:
in each attempt below, please add short comment on what makes that attempt wrong (e.g. wrong name, wrong args, etc)